### DISCUSSION TITLE
VO2 Max take time?

### DISCUSSION THREAD

**Discussion by jheizer** [philosowaffle/peloton-to-garmin#229] [Q&A]
Do you know if it takes a certain number of rides before the VO2 Max started to work for you or was it immediate?  I updated the other day and today made sure to stay above the 70% HR mark for over 20min.  I wasn't sure if you noticed it takes a few rides or worked immediately.   It did upload as Tacx App (Windows)

![image](https://user-images.githubusercontent.com/1015493/155244168-2a818377-f7e1-4dd0-82ee-9059b32c2e94.png)

![image](https://user-images.githubusercontent.com/1015493/155244200-c14a9247-1ecb-4059-ba7d-b2ab4734a478.png)

![image](https://user-images.githubusercontent.com/1015493/155244324-7fdc4135-ed31-4078-8d30-d416cdda22fb.png)
It may look close to the 128, but its always 133 or above skimming along the line.

**Comment by philosowaffle** [comment]
For me it showed up immediately after I synced a ride that:
1) Set the device info on the workout to be TACX
2) Met all of the Garmin criteria you mentioned above

**Reply by jheizer** [reply]
I tried again and no luck. I guess it doesn't like me. 3:30 on I made sure to stay above 132bpm till the end.
Legs were pretty beat so it was all I could muster today. https://connect.garmin.com/modern/activity/8347451333

**Reply by philosowaffle** [reply]
I'm disappointed its not working for you. We were only able to verify it worked for 2 different people which is admittedly not a large sample size.

In the Reports section, are you getting other data on the TrainingStatus charts?

**Reply by jheizer** [reply]
Hmmm, maybe its a by product of the watch I have. I have a vivoactive 4. I don't have a Training Status chart option at all. I have VO2Max for runs last fall, and all of my biking logs were mountain biking where I'm not pushing the heart rate to these limits. Skimming my watch's feature list is specifically says under "Training, Planning and Analysis Features" VO2 MAX (RUN) so I wonder if its not available for biking and in turn not working for this just like you found with having no watch. No watch or not a good enough one we lack the "license key" to the features.

**Reply by philosowaffle** [reply]
> No watch or not a good enough one we lack the "license key" to the features.

This does seem to be the case. Its amazing to me that simply changing the Device Type in the FIT file, which is purely a metadata change, suddenly "unlocks" access to VO2 Max (for some of us).  They clearly have all the data they need to calculate a VO2 Max from these workout files, they simply choose not to.

**Reply by Vicore1** [reply]
Here is one a spun up (spinned up?) [](https://connect.garmin.com/modern/activity/8353980255) as you can see I way over 70% for more then 20 mins.  Just like Jheizer I don't have a "Training Status". I have the same default zones as heart rates as he does. Also, I am not getting credit toward the current Tracx badge either, can we confirm that everyone but us is getting credit toward that badge, it might be the same backend flag that enables that also?
 I am wondering if contacting Garmin support playing dumb as to why my Tracx is not recording correctly they can flip whatever flag enables this functionality?

**Reply by Vicore1** [reply]
(https://connect.garmin.com/modern/activity/8353980255)

**Reply by philosowaffle** [reply]
Confirmed on my end its not counting towards the Tacx badges.

**Reply by Vicore1** [reply]
I did do another ride that should have counted last night after downloading the app on my phone and attaching it to my Garmin account :https://connect.garmin.com/modern/activity/8354533908 it did not generate the additional metrics

**Comment by Vicore1** [comment]
I have the same watch as my default device in Garmin. I have had a sick kid at home and haven't had the time to get a good ride in the last few days.  Can you confirm requirements to get get a ride to count toward V02Max and I will knock one out later today and see what happens.

**Reply by philosowaffle** [reply]
1. Be running version 2.2.0 of P2G
2. Remove any custom device info in your configuration so that P2G will upload as the new default device TACX
3. Meet all of [Garmins requirements](https://support.garmin.com/en-SG/?faq=MyIZ05OMpu6wSl95UVUjp7)

**Reply by Vicore1** [reply]
So as we can see from my reply to jheizer it didn't work (at least not yet).  But me thinks I might be able to figure it out.
 I am going to continue to try some things:
Firstly, I went and created a TACX account( It is free so no harm in trying that, I would even be will to pay for it for a month to see if that makes any difference). 
- I did this as an attempt to turn on whatever flag isn't currently enabled
- My guess is the flag for the TACX and those other baller watches is the same and it isn't turned on my account.  
I believe this to be the case because I have done multiple rides with a TACX device (my peloton rides) and I am not getting credit for the badges.
Second, looking at this article: https://support.garmin.com/en-US/?faq=apva7Th82K6Mm0zY4shzD6
I am going to take my watch off completely and use a heart rate strap.

I already rode for an hour today, let me see if have another half way decent 30mins in me try this out.

**Reply by Vicore1** [reply]
Also, this article seems to add some additional requirements that aren't mentioned because frankly I don't think they planned for people to upload Tacx stuff without tacx accounts:
"For your Training Status to update from a supported third-party app, you must also meet these requirements:
The third-party app must be linked to your Garmin Connect account.
You must have a Training Status capable device that you are actively recording running or cycling activities with. If you remove your compatible device from Garmin Connect, you will not receive Training Status data from a third-party or Tacx app activity. 
Physio TrueUp must be enabled on your Garmin device (see  [What is Physio TrueUp?](https://support.garmin.com/en-US/?faq=EjPECQK58qA0xzJ5X74vm7))."
https://support.garmin.com/en-US/?faq=VxKazDQ2mkAmDoQbJriEBA

PS our watch does support Physio TrueUp so that is a plus

**Reply by philosowaffle** [reply]
Garmin does claim to support the following devices as well for cycling VO2 max:

* epix
* fenix 3
* quatix 3
* tactix bravo

Folks could try switching to Fenix 3 or an Epix device. That would eliminate any 3rd party app complexities.

@tejchicago Would you be willing to share your Epix device info file (you can zero out the unit ID before sharing)?  You can find instructions [here](https://philosowaffle.github.io/peloton-to-garmin/configuration/providing-device-info.html#steps). This would let others experiment with using an Epix device instead of Tacx to see if that makes any difference.

**Reply by Vicore1** [reply]
I have a feeling that it might not make a difference unless you have one linked to your garmin account ( am guessing the device on the account is what unlocks the functionality), but it is worth a try. I am not sure I am willing to drop the cash on another watch just to maybe get these metrics let alone for testing something Garmin could figure out and patch out.

**Reply by tejchicago** [reply]
Sure, here's the device info for my watch, (will have to rename it to an XML file) hope that helps. But to be clear, my Peloton workouts are showing up from Tacx, I wasn't getting cycling V02 Max data from just the watch.

[EpixGen2 - deviceInfo.txt](https://github.com/philosowaffle/peloton-to-garmin/files/8142483/EpixGen2.-.deviceInfo.txt)
!

**Reply by Vicore1** [reply]
tejchicago thank you so much helping out with this.  
Firstly, can you confirm you are or are not recieving creidt for the TacX badges ( I am just curious) 
So I took this and ftp into my container as an xml and confirmed my json has  "DeviceInfoPath": "./deviceInfo.xml".  I am going recreate my container.  Other then that is there anything I am missing other then riding one out as a test?

**Reply by philosowaffle** [reply]
Re-creating the container may delete the file you added.  Instead, I would "mount" the file into the container the exact same way you mount the configuration file.  Then recreate the container. This will ensure the file persists across container restarts.

If you're using docker-compose it would look something like this:
```
version: "3.9"
services:
  p2g:
    container_name: p2g
    image: philosowaffle/peloton-to-garmin:stable
    environment:
      - TZ=America/Chicago
    volumes:
      - ./configuration.local.json:/app/configuration.local.json
      - ./output:/app/output
      - ./deviceInfo.xml:/app/deviceInfo.xml
```

And if you're using docker run then I think the command is something like:
```
docker run  philosowaffle/peloton-to-garmin:stable -v ./configuration.local.json:/app/configuration.local.json -v ./deviceInfo.xml:/app/deviceInfo.xml
```

**Reply by Vicore1** [reply]
I have previously been just using `docker-compose up --detach` in the location where the json and the yaml were located. I updated the compose file to include the  - ./deviceInfo.xml:/app/deviceInfo.xml and I added the deviceInfo.xml into that that same parent directory as those two other files.  I just did a few stretches and a 10min bike to confirm I didn't jack anything up and now I am getting no device information listed on the workouts. 

Here is my previous compose:
version: "3.5"

services:
  p2g:
    container_name: p2g
    image: philosowaffle/peloton-to-garmin:stable
    environment:
      - TZ=America/Chicago
    volumes:
      - ./configuration.local.json:/app/configuration.local.json
      - ./output:/app/output
      - ./working:/app/working

Here is what I changed it to:
version: "3.9"
services:
  p2g:
    container_name: p2g
    image: philosowaffle/peloton-to-garmin:stable
    environment:
      - TZ=America/Chicago
    volumes:
      - ./configuration.local.json:/app/configuration.local.json
      - ./output:/app/output
      - ./working:/app/working
      - ./deviceInfo.xml:/app/deviceInfo.xml

oh and is what my output log is looking like (after it successfully imported the exercises w/ no device):

2022-02-25 12:15:37.797 -06:00 [INF] Sleeping for 7200 seconds...
2022-02-25 12:25:30.368 -06:00 [WRN] App.PythonAndGuploadInstalledLocally setting is deprecated and will be removed in a future release. Please swith to using Garmin.Up
loadStrategy config.
2022-02-25 12:25:30.706 -06:00 [WRN] App.PythonAndGuploadInstalledLocally setting is deprecated and will be removed in a future release. Please swith to using Garmin.Up
loadStrategy config.
2022-02-25 12:25:30.782 -06:00 [ERR] Failed to get syncServiceStatus from db
System.Collections.Generic.KeyNotFoundException: The given key was not present in the dictionary.
   at JsonFlatFileDataStore.DataStore.GetItem[T](String key)
   at Common.Database.SyncStatusDb.GetSyncStatusAsync() in /build/src/Common/Database/SyncStatusDb.cs:line 38
2022-02-25 12:25:31.488 -06:00 [INF] Application started. Press Ctrl+C to shut down.
2022-02-25 12:25:31.494 -06:00 [INF] Hosting environment: Production
2022-02-25 12:25:31.499 -06:00 [INF] Content root path: /app
2022-02-25 12:25:32.656 -06:00 [INF] No files to convert in download directory. Nothing to do. ./working/downloaded
2022-02-25 12:25:32.664 -06:00 [INF] No upload directory found. Nothing to do.
2022-02-25 12:25:32.666 -06:00 [ERR] Failed to clean up working directory: ./working
System.IO.IOException: Device or resource busy
   at System.IO.FileSystem.RemoveDirectoryInternal(DirectoryInfo directory, Boolean recursive, Boolean throwOnTopLevelDirectoryNotFound)
   at System.IO.FileSystem.RemoveDirectory(String fullPath, Boolean recursive)
   at System.IO.Directory.Delete(String path, Boolean recursive)
   at Common.IOWrapper.Cleanup(String dir) in /build/src/Common/FileHandling.cs:line 114
2022-02-25 12:25:32.795 -06:00 [INF] Sleeping for 7200 seconds...

**Reply by philosowaffle** [reply]
@Vicore1 Your compose file changes look good. In the `deviceInfo.xml` can you try adding a value for `UnitId`?  Any random number should do.

**Reply by Vicore1** [reply]
will that become the default device for everything that comes from peloton or just cycling ( assuming I can get it to work)? asking because I would like to troubleshoot it without having get on the bike :)

**Reply by philosowaffle** [reply]
yup, default for all workout types, so you can do a meditation :)   Appreciate you testing these things out for folks!

**Reply by tejchicago** [reply]
FYI - I'm not seeing any Tacx badges (which I didn't realize were a thing), I have plenty of other badges.

I did notice today that my non-cycling Peloton activities (strength and stretching) now show they're coming from an Edge 1000 device.
Doesn't matter in the least for me, just happy they get captured.

**Reply by Vicore1** [reply]
I noticed that in the newest version before I started messing w/ the deviceInfo.xml, so it isn't just you.

**Reply by Vicore1** [reply]
For the life of me I can't seem to get the deviceInfo.xml to work as expected, per your request I updated the UnitId:
 $ more deviceInfo.xml
<Creator>
        <Name>EPIX Gen2</Name>
        <UnitId>86753</UnitId>
        <ProductID>3943</ProductID>
        <Version>
        <VersionMajor>7</VersionMajor>
        <VersionMinor>24</VersionMinor>
        <BuildMajor>0</BuildMajor>
        <BuildMinor>0</BuildMinor>
        </Version>
</Creator>
 
I am now getting the outline of a Garmin watch on the website(still getting nothin in the app for the device) where prior to that I wasn't getting anything
w/ UnitId
https://connect.garmin.com/modern/activity/8359334335
w/o UnitId
Here is a previous one:
https://connect.garmin.com/modern/activity/8357827259

Also on another note, is "- ./working:/app/working" no longer needed? I looked at the wiki and for giggles I tried to spin up a totally new container using the latest stable, the example config and example compose and it errored out all over the place, it appears I am doing something wrong. I am going to give in fresh eyes sometime tomorrow.

**Reply by philosowaffle** [reply]
I tested this out today and confirmed the following:

1. The Epix device info is not considered valid by Garmin for some reason 🤷 
2. For non-cycling workouts Connect is showing an Edge device which is unexpected, created: #232

**Reply by gbassani** [reply]
Has anyone confirmed if the Epix2 is still 'not valid', or is it working now and providing VO2 max data? If unknown, I can try re-configuring my setup this weekend to an epix profile.

**Comment by tejchicago** [comment]
FWIW- I am getting VO2 Max cycling updates with this version (after 6 years on the Peloton)!   Brand new shiny watch, Epix Gen 2.  I now see 2 VO2 Max readings, one for running, one for cycling.   

One thought/question - really a Garmin one, not P2G.  I've always started an indoor cycling activity on my watch, use the HR monitor to broadcast to the bike.  At the end of my ride, I stop the workout activity and discard it on the watch.   P2G later picks it up and feeds it to Garmin.

I'm wondering if I'm now duplicating any input to Garmin with this approach?  Nothing in the workout metrics, but overall for HR maybe?

In any case, nice work @philosowaffle , this is outstanding.  Now I just need to improve my numbers!

**Reply by Vicore1** [reply]
For us vivoactive 4 folks that isn't even an option, we can't broadcast the HR to the bike as far as I know. I use a chest strap, I don't even have the watch on most when I am on the bike most of the time.

**Reply by jheizer** [reply]
Yes we can on vivoactive 4.  Press and hold bottom button, settings, Wrist Heart Rate, Broadcast Off and accept turning it on.  HR will show constantly on the screen then back button to kill it when done.  Works pretty good.

**Reply by Vicore1** [reply]
cool I will try it, thanks for the heads up

**Reply by philosowaffle** [reply]
@tejchicago To the best of my knowledge, as long as you choose `Discard` at the end of the workout, the data is never sent to Garmin so there will not be any duplicate data.

**Comment by danielsfrazier** [comment]
I changed from the windows app to android app in an attempt to get
the rides to count towards badges which is not working for me. I have the
tacx app on my android and linked to my garmin account but the rides don't
show up in the tacx app. Perhaps they need to be recorded by the app and
not uploaded in order to count towards badges.

I checked in my garmin account and I don't have zwift, trainerroad or etc.
linked to my account. I'm still able to get vo2 and training status using
the following device info. My primary watch is a FR945. If I recall
correctly, I originally 'unlocked' all the advanced training metrics when I
upgraded from a FR235 to a FR645. I only have one activity from my FR945
that generated vo2 while biking using a HRM, speed sensor 2, cadence sensor
and FB3's 'Dozen Cycle' iq app. I don't have a power meter on my bike so
the iq app estimated power. All other cycle vo2 values have been added by
the p2g integration by changing to Tacx device type.

<Creator>
  <Name>TacxTrainingAppAndroid</Name>
  <UnitId>8675309</UnitId>
  <ProductID>30045</ProductID>
  <Version>
    <VersionMajor>4</VersionMajor>
    <VersionMinor>26</VersionMinor>
    <BuildMajor>1</BuildMajor>
    <BuildMinor>0</BuildMinor>
  </Version>
</Creator>

**Reply by Vicore1** [reply]
I tried that last night and no dice.  I installed the desktop app and was going to it also at some point even know I don't think it is going to work just to prove out it doesn't.

**Reply by philosowaffle** [reply]
At this point, it unfortunately seems pretty definitive that Garmin has locked enhanced training metrics behind physically owning a certain type of watch. :(

I will atleast update the P2G documentation to reflect this so others know when they can expect to receive VO2 max data.

**Comment by Vicore1** [comment]
Yup, I agree.
